File: packages/template-python/src/deepagents_template/runtime/config/config_sources.py
"""Config sources — load/config overlays from files, env, and plugins."""

from __future__ import annotations

import contextlib
import json
import os
from pathlib import Path
from typing import Any

from deepagents_template.runtime.config.config_merge import setNestedValue
from deepagents_template.runtime.config.config_paths import resolveConfigResourcePath, resolvePath
from deepagents_template.runtime.logger import logger


def loadJsonFile(filePath: str, baseDir: str | None = None) -> dict[str, Any] | None:
    resolved = resolvePath(filePath, baseDir or os.getcwd())
    p = Path(resolved)
    if not p.exists():
        return None
    try:
        return json.loads(p.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError) as exc:
        logger.warning(f"Failed to parse config file: {resolved}", extra={"error": str(exc)})
        return None


def normalizeConfigResourcePaths(config: dict[str, Any], baseDir: str) -> None:
    mcp = config.get("mcp")
    if isinstance(mcp, dict):
        if isinstance(mcp.get("configPath"), str):
            mcp["configPath"] = resolveConfigResourcePath(mcp["configPath"], baseDir)
        if isinstance(mcp.get("configPaths"), list):
            mcp["configPaths"] = [
                resolveConfigResourcePath(e, baseDir) if isinstance(e, str) else e
                for e in mcp["configPaths"]
            ]
    skills = config.get("skills")
    if isinstance(skills, dict) and isinstance(skills.get("directories"), list):
        skills["directories"] = [
            resolveConfigResourcePath(e, baseDir) if isinstance(e, str) else e
            for e in skills["directories"]
        ]
    if isinstance(config.get("agentsDirectories"), list):
        config["agentsDirectories"] = [
            resolveConfigResourcePath(e, baseDir) if isinstance(e, str) else e
            for e in config["agentsDirectories"]
        ]
    memory = config.get("memory")
    if isinstance(memory, dict) and isinstance(memory.get("dir"), str):
        memory["dir"] = resolveConfigResourcePath(memory["dir"], baseDir)
    agent = config.get("agent")
    if isinstance(agent, dict) and isinstance(agent.get("systemPromptPath"), str):
        agent["systemPromptPath"] = resolveConfigResourcePath(agent["systemPromptPath"], baseDir)


def loadModelsOverlayFromFile(filePath: str) -> dict[str, Any] | None:
    raw = loadJsonFile(filePath)
    if raw is None:
        return None
    if "model" in raw and isinstance(raw["model"], dict):
        return raw
    if "default" in raw and isinstance(raw["default"], dict):
        return {"model": raw["default"]}
    return {"model": raw}


def loadMcpOverlayFromFile(filePath: str) -> dict[str, Any] | None:
    p = Path(resolvePath(filePath))
    if not p.exists():
        return None
    return {"mcp": {"configPaths": [str(p)]}}


ENV_MAP: dict[str, str] = {
    "ACP_AGENT_NAME": "agent.name",
    "ACP_AGENT_DESCRIPTION": "agent.description",
    "AGENT_SYSTEM_PROMPT": "agent.systemPrompt",
    "AGENT_SYSTEM_PROMPT_PATH": "agent.systemPromptPath",
    "DEEPAGENTS_WORKING_DIR": "workspace.workingDir",
    "AGENT_WORKING_DIR": "workspace.workingDir",
    "PLATFORM_API_BASE_URL": "platform.apiBaseUrl",
    "PLATFORM_AGENT_ID": "platform.agentId",
    "PLATFORM_SPACE_ID": "platform.spaceId",
    "DEFAULT_MODEL": "model.name",
    "ANTHROPIC_MODEL": "model.name",
    "ANTHROPIC_BASE_URL": "model.baseUrl",
    "OPENAI_MODEL": "model.name",
    "OPENAI_BASE_URL": "model.baseUrl",
    "LLM_PROVIDER": "model.provider",
    "MAX_TOKENS": "model.settings.maxTokens",
    "MCP_CONFIG_PATH": "mcp.configPath",
    "LOG_LEVEL": "logging.level",
    "DEEPAGENTS_PERMISSIONS_MODE": "permissions.mode",
    "DEEPAGENTS_SANDBOX_PROFILE": "sandbox.profile",
}

NUMERIC_ENV_KEYS: set[str] = {"MAX_TOKENS"}


def loadFromEnv() -> dict[str, Any]:
    overlay: dict[str, Any] = {}
    for env_key, config_path in ENV_MAP.items():
        value = os.environ.get(env_key)
        if value and value != "":
            if env_key in NUMERIC_ENV_KEYS:
                with contextlib.suppress(ValueError):
                    setNestedValue(overlay, config_path, int(value))
            else:
                setNestedValue(overlay, config_path, value)
    acp_debug = os.environ.get("ACP_DEBUG")
    if acp_debug in ("true", "1"):
        setNestedValue(overlay, "logging.level", "debug")
    return overlay


def inferModelProviderIfUnset(config: dict[str, Any]) -> dict[str, Any]:
    explicit = (os.environ.get("LLM_PROVIDER") or "").strip().lower()
    if explicit in ("openai", "anthropic"):
        return config
    has_openai = bool(
        os.environ.get("OPENAI_API_KEY", "").strip()
        or os.environ.get("OPENAI_BASE_URL", "").strip()
    )
    has_anthropic = bool(
        os.environ.get("ANTHROPIC_API_KEY", "").strip()
        or os.environ.get("ANTHROPIC_AUTH_TOKEN", "").strip()
        or os.environ.get("ANTHROPIC_BASE_URL", "").strip()
    )
    provider: str | None = None
    if has_openai and not has_anthropic:
        provider = "openai"
    elif has_anthropic and not has_openai:
        provider = "anthropic"
    elif has_openai and has_anthropic and os.environ.get("OPENAI_API_KEY", "").strip():
        provider = "openai"
    if not provider or provider == config.get("model", {}).get("provider"):
        return config
    model = dict(config.get("model", {}))
    model["provider"] = provider
    if provider == "openai":
        model["apiKeyEnv"] = "OPENAI_API_KEY"
    result = dict(config)
    result["model"] = model
    return result


def discoverPluginManifests(pluginRoot: str) -> list[str]:
    root = Path(pluginRoot)
    if not root.exists():
        return []
    root_manifest = root / "plugin.json"
    if root_manifest.exists():
        return [str(root_manifest)]
    results: list[str] = []
    for entry in sorted(root.iterdir()):
        if entry.is_dir():
            mf = entry / "plugin.json"
            if mf.exists():
                results.append(str(mf))
    return results


def loadPluginOverlay(config: dict[str, Any], workspaceRoot: str) -> dict[str, Any] | None:
    overlay: dict[str, Any] = {}
    enabled_set = set(config.get("plugins", {}).get("enabled", []))
    disabled_set = set(config.get("plugins", {}).get("disabled", []))
    for plugin_dir in config.get("plugins", {}).get("directories", []):
        resolved_dir = resolvePath(plugin_dir, workspaceRoot)
        for manifest_path in discoverPluginManifests(resolved_dir):
            manifest = loadJsonFile(manifest_path)
            if not manifest:
                continue
            plugin_id = manifest.get("id") or manifest.get("name") or manifest_path
            if manifest.get("enabled") is False or plugin_id in disabled_set:
                continue
            if enabled_set and plugin_id not in enabled_set and "*" not in enabled_set:
                continue
            base_dir = str(Path(manifest_path).parent)
            skills_dirs = manifest.get("skillsDirectories", [])
            if skills_dirs:
                existing = overlay.setdefault("skills", {}).setdefault("directories", [])
                existing.extend(resolvePath(p, base_dir) for p in skills_dirs)
            agents_dirs = manifest.get("agentsDirectories", [])
            if agents_dirs:
                existing = overlay.setdefault("agentsDirectories", [])
                existing.extend(resolvePath(p, base_dir) for p in agents_dirs)
            hooks = manifest.get("hooks", [])
            if hooks:
                overlay.setdefault("hooks", []).extend(hooks)
            mcp = manifest.get("mcp", {}) or {}
            mcp_servers = manifest.get("mcpServers", {}) or {}
            config_paths = list(mcp.get("configPaths", []))
            if mcp.get("configPath"):
                config_paths.append(mcp["configPath"])
            if config_paths or mcp_servers:
                mcp_overlay = overlay.setdefault("mcp", {})
                mcp_overlay.setdefault("configPaths", []).extend(
                    resolvePath(p, base_dir) for p in config_paths
                )
                existing_servers = mcp_overlay.setdefault("servers", {})
                existing_servers.update(mcp.get("servers", {}))
                existing_servers.update(mcp_servers)
    return overlay if overlay else None

File: packages/template/src/runtime/config/config-sources.ts
/**
 * Config sources.
 *
 * Load and shape overlays from the inputs that feed the loader — JSON config
 * files, plugin manifests, and environment variables. Extracted from
 * config-loader.ts; the orchestration that combines these lives there.
 */
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { logger } from "../logger.js";
import type { AppConfig } from "./config-schema.js";
import { resolvePath, resolveConfigResourcePath } from "./config-paths.js";
import { setNestedValue, isRecord } from "./config-merge.js";

/** Load JSON config from file path */
export function loadJsonFile(filePath: string, baseDir = process.cwd()): Record<string, unknown> | null {
  const resolved = resolvePath(filePath, baseDir);
  if (!existsSync(resolved)) {
    return null;
  }
  try {
    const content = readFileSync(resolved, "utf-8");
    return JSON.parse(content) as Record<string, unknown>;
  } catch (err) {
    logger.warn(`Failed to parse config file: ${resolved}`, {
      error: err instanceof Error ? err.message : String(err),
    });
    return null;
  }
}

export function normalizeConfigResourcePaths(config: Record<string, unknown>, baseDir: string): void {
  const mcp = config.mcp;
  if (isRecord(mcp)) {
    if (typeof mcp.configPath === "string") {
      mcp.configPath = resolveConfigResourcePath(mcp.configPath, baseDir);
    }
    if (Array.isArray(mcp.configPaths)) {
      mcp.configPaths = mcp.configPaths.map((entry: unknown) =>
        typeof entry === "string" ? resolveConfigResourcePath(entry, baseDir) : entry
      );
    }
  }

  const skills = config.skills;
  if (isRecord(skills) && Array.isArray(skills.directories)) {
    skills.directories = skills.directories.map((entry: unknown) =>
      typeof entry === "string" ? resolveConfigResourcePath(entry, baseDir) : entry
    );
  }

  if (Array.isArray(config.agentsDirectories)) {
    config.agentsDirectories = config.agentsDirectories.map((entry: unknown) =>
      typeof entry === "string" ? resolveConfigResourcePath(entry, baseDir) : entry
    );
  }

  const memory = config.memory;
  if (isRecord(memory) && typeof memory.dir === "string") {
    memory.dir = resolveConfigResourcePath(memory.dir, baseDir);
  }

  const agent = config.agent;
  if (isRecord(agent) && typeof agent.systemPromptPath === "string") {
    agent.systemPromptPath = resolveConfigResourcePath(agent.systemPromptPath, baseDir);
  }
}

export function modelsOverlayFromFile(filePath: string): Partial<AppConfig> | null {
  const raw = loadJsonFile(filePath);
  if (!raw) {
    return null;
  }
  if (raw.model && typeof raw.model === "object") {
    return raw as Partial<AppConfig>;
  }
  if (raw.default && typeof raw.default === "object") {
    return { model: raw.default as AppConfig["model"] } as Partial<AppConfig>;
  }
  return { model: raw as AppConfig["model"] } as Partial<AppConfig>;
}

export function mcpOverlayFromFile(filePath: string): Partial<AppConfig> | null {
  if (!existsSync(resolvePath(filePath))) {
    return null;
  }
  return { mcp: { configPaths: [resolvePath(filePath)] } as AppConfig["mcp"] } as Partial<AppConfig>;
}

interface PluginManifest {
  id?: string;
  name?: string;
  enabled?: boolean;
  skillsDirectories?: string[];
  agentsDirectories?: string[];
  hooks?: AppConfig["hooks"];
  mcp?: {
    configPath?: string;
    configPaths?: string[];
    servers?: Record<string, unknown>;
  };
  mcpServers?: Record<string, unknown>;
}

export function loadPluginOverlay(config: AppConfig, workspaceRoot: string): Partial<AppConfig> | null {
  const overlay: Partial<AppConfig> = {};
  const enabled = new Set(config.plugins.enabled);
  const disabled = new Set(config.plugins.disabled);

  for (const pluginDir of config.plugins.directories) {
    const resolvedDir = resolvePath(pluginDir, workspaceRoot);
    for (const manifestPath of discoverPluginManifests(resolvedDir)) {
      const manifest = loadJsonFile(manifestPath) as PluginManifest | null;
      if (!manifest) {
        continue;
      }
      const pluginId = manifest.id || manifest.name || manifestPath;
      if (manifest.enabled === false || disabled.has(pluginId)) {
        continue;
      }
      if (enabled.size > 0 && !enabled.has(pluginId)) {
        continue;
      }

      const baseDir = dirname(manifestPath);
      if (manifest.skillsDirectories?.length) {
        overlay.skills = {
          ...overlay.skills,
          directories: [
            ...(overlay.skills?.directories ?? []),
            ...manifest.skillsDirectories.map((path) => resolvePath(path, baseDir)),
          ],
        } as Partial<AppConfig["skills"]> as AppConfig["skills"];
      }
      if (manifest.agentsDirectories?.length) {
        overlay.agentsDirectories = [
          ...(overlay.agentsDirectories ?? []),
          ...manifest.agentsDirectories.map((path) => resolvePath(path, baseDir)),
        ];
      }
      if (manifest.hooks?.length) {
        overlay.hooks = [...(overlay.hooks ?? []), ...manifest.hooks];
      }

      const configPaths = [
        ...(manifest.mcp?.configPath ? [manifest.mcp.configPath] : []),
        ...(manifest.mcp?.configPaths ?? []),
      ].map((path) => resolvePath(path, baseDir));
      const servers = {
        ...(manifest.mcp?.servers ?? {}),
        ...(manifest.mcpServers ?? {}),
      };
      if (configPaths.length > 0 || Object.keys(servers).length > 0) {
        overlay.mcp = {
          ...overlay.mcp,
          configPaths: [...(overlay.mcp?.configPaths ?? []), ...configPaths],
          servers: { ...(overlay.mcp?.servers ?? {}), ...servers },
        } as AppConfig["mcp"];
      }
    }
  }

  return Object.keys(overlay).length > 0 ? overlay : null;
}

function discoverPluginManifests(pluginRoot: string): string[] {
  if (!existsSync(pluginRoot)) {
    return [];
  }
  const rootManifest = resolve(pluginRoot, "plugin.json");
  if (existsSync(rootManifest)) {
    return [rootManifest];
  }
  return readdirSync(pluginRoot)
    .map((entry) => resolve(pluginRoot, entry))
    .filter((entryPath) => {
      try {
        return statSync(entryPath).isDirectory();
      } catch {
        return false;
      }
    })
    .map((entryPath) => resolve(entryPath, "plugin.json"))
    .filter((manifestPath) => existsSync(manifestPath));
}

// ─── Environment Variable Mapping ───────────────────────

const ENV_MAP: Record<string, string> = {
  ACP_AGENT_NAME: "agent.name",
  ACP_AGENT_DESCRIPTION: "agent.description",
  AGENT_SYSTEM_PROMPT: "agent.systemPrompt",
  AGENT_SYSTEM_PROMPT_PATH: "agent.systemPromptPath",
  DEEPAGENTS_WORKING_DIR: "workspace.workingDir",
  AGENT_WORKING_DIR: "workspace.workingDir",
  PLATFORM_API_BASE_URL: "platform.apiBaseUrl",
  PLATFORM_AGENT_ID: "platform.agentId",
  PLATFORM_SPACE_ID: "platform.spaceId",
  DEFAULT_MODEL: "model.name",
  ANTHROPIC_MODEL: "model.name",
  ANTHROPIC_BASE_URL: "model.baseUrl",
  OPENAI_MODEL: "model.name",
  OPENAI_BASE_URL: "model.baseUrl",
  LLM_PROVIDER: "model.provider",
  MAX_TOKENS: "model.settings.maxTokens",
  MCP_CONFIG_PATH: "mcp.configPath",
  LOG_LEVEL: "logging.level",
  DEEPAGENTS_PERMISSIONS_MODE: "permissions.mode",
  DEEPAGENTS_SANDBOX_PROFILE: "sandbox.profile",
};

/** Numeric env vars that need parsing */
const NUMERIC_ENV_KEYS = new Set(["MAX_TOKENS"]);

/** Build config overlay from environment variables */
export function loadFromEnv(): Partial<AppConfig> {
  const overlay: Record<string, unknown> = {};
  for (const [envKey, configPath] of Object.entries(ENV_MAP)) {
    const value = process.env[envKey];
    if (value !== undefined && value !== "") {
      // Parse numeric env vars
      if (NUMERIC_ENV_KEYS.has(envKey)) {
        const num = Number(value);
        if (!isNaN(num)) {
          setNestedValue(overlay, configPath, num);
        }
      } else {
        setNestedValue(overlay, configPath, value);
      }
    }
  }
  // ACP_DEBUG is a boolean flag that maps to logging.level = "debug"
  const acpDebug = process.env.ACP_DEBUG;
  if (acpDebug === "true" || acpDebug === "1") {
    setNestedValue(overlay, "logging.level", "debug");
  }
  return overlay as Partial<AppConfig>;
}

/**
 * 当未设置 LLM_PROVIDER 时，根据环境变量中的凭证族自动推断 model.provider。
 *
 * Zed / rcoder 的 OpenAI 兼容 profile 通常只注入 OPENAI_*，不应再强制要求
 * LLM_PROVIDER=openai。显式设置 LLM_PROVIDER 时仍以该值为准。
 */
export function inferModelProviderIfUnset(config: AppConfig): AppConfig {
  const explicit = process.env.LLM_PROVIDER?.trim().toLowerCase();
  if (explicit === "openai" || explicit === "anthropic") {
    return config;
  }

  const hasOpenAISignals = !!(
    process.env.OPENAI_API_KEY?.trim() ||
    process.env.OPENAI_BASE_URL?.trim()
  );
  const hasAnthropicSignals = !!(
    process.env.ANTHROPIC_API_KEY?.trim() ||
    process.env.ANTHROPIC_AUTH_TOKEN?.trim() ||
    process.env.ANTHROPIC_BASE_URL?.trim()
  );

  let provider: AppConfig["model"]["provider"] | null = null;

  if (hasOpenAISignals && !hasAnthropicSignals) {
    provider = "openai";
  } else if (hasAnthropicSignals && !hasOpenAISignals) {
    provider = "anthropic";
  } else if (hasOpenAISignals && hasAnthropicSignals && process.env.OPENAI_API_KEY?.trim()) {
    // 两套凭证同时存在且未指定 LLM_PROVIDER：有 OPENAI_API_KEY 时走 OpenAI 兼容路径
    provider = "openai";
  }

  if (!provider || provider === config.model.provider) {
    return config;
  }

  return {
    ...config,
    model: {
      ...config.model,
      provider,
      // OpenAI 路径默认从 OPENAI_API_KEY 取密钥
      ...(provider === "openai" ? { apiKeyEnv: "OPENAI_API_KEY" } : {}),
    },
  };
}
